CONTRIBUTING.rst

```diff
@@ -380,6 +380,56 @@ pull requests from other contributors yourself after having reviewed
 them.
 
 
+Merge/squash guidelines
+-----------------------
+
+When a PR is approved and ready to be integrated to the `main` branch, one has the option to *merge* the commits unchanged, or *squash* all the commits into a single commit.
+
+Here is some guidelines on how to proceeed, based on the state of the PR commit history:
+
+1. Miscellaneous commits:
+
+   * `Implement X`
+   * `Fix test_a`
+   * `Add myself to AUTHORS`
+   * `!fixup Fix test_a`
+   * `Update tests/test_integration.py`
+   * `Update tests/test_integration.py`
+
+   In this case, prefer to use the **Squash** merge strategy: the commit history is a bit messy (not in a derrogatory way, often one just commits changes because they know the changes will eventually be squashed together), so squashing everything into a single commit is best. Prefer to also edit the default GitHub message to add more details and improve on it.
+
+2. Separate commits related to the same topic:
+
+   * `Implement X`
+   * `Add myself to AUTHORS`
+   * `Update CHANGELOG for X`
+
+   In this case, prefer to use the **Squash** merge strategy: while the commit history is not "messy" as in the example above, the individual commits do not bring much value overall, specially when looking at the changes a few months/years down the line.
```

Comment:
How about “Add tests for X”? Such commits totally deserve to be revertible separately...

Reply:
Sorry but again we disagree: I don't see the point of keeping the tests if we are reverting the feature? And vice versa. And even if/when that happens (which TBH I have never seen happen in pytest's entire history) I guess we can cherry-pick/edit manually -- but given this is really rare, I don't think we need an exception/rule for this.

Reply:
This is quite a controversial example overall. It's not only about the tests (which might need to be reverted because they were relying on incorrect expectations, for example). Even with adding oneself to the authors. Imagine removing the entire thing, but a person contributed more things after that first PR, and you'd just be deleting their credit while it would be better to keep it. Same with the change notes. Although, with these, I'd argue that they should never be reverted even if the feature is, since the contribution has been made even if its code has been modified/rewritten/removed at some point. The same goes for changelog entries — if something was implemented in 8.2 and reverted in 8.3, it doesn't mean that the change log should eliminate any mentions of it from historic documents.

Reply:
>  Even with adding oneself to the authors. Imagine removing the entire thing, but a person contributed more things after that first PR, and you'd just be deleting their credit while it would be better to keep it.

Contributions are kept (unless someone explicitly removes them) via the `Co-authored-by` field, for an example:

https://github.com/pytest-dev/pytest/commit/d489247505a953885a156e61d4473497cbc167ea


I understand your examples, but they seem to me more an exception rather than something which happens all the time and hence we should always merge (instead of squashing).

Reply:
I'm talking about the `AUTHORS` file, which is what that `Add myself to AUTHORS` example commit message implies. Git commits are of course kept in history, I know that.

And yes, my point is that natural merge is better here.

Reply:
Understand your point, but I still think this is not a good approach because these reverts we are discussing is incredibly rare in my experience (both in pytest and in other repositories).

Reply:
Perhaps. Though, being rare, it makes it harder for people to remember to make partial reverts.